src/common/framework/params_builder.ts

```diff
@@ -1,15 +1,98 @@
 import { CaseParams, CaseParamsIterable, publicParamsEquals } from './params_utils.js';
 import { assert } from './util/util.js';
 
-// https://stackoverflow.com/a/56375136
+/** Forces a type to resolve its type definitions, to make it readable/debuggable. */
+export type ResolveType<T> = T extends object
+  ? T extends infer O
+    ? { [K in keyof O]: ResolveType<O[K]> }
+    : never
+  : T;
+
+/**
+ * Computes the intersection of a set of types, given the union of those types.
+ *
+ * From: https://stackoverflow.com/a/56375136
+ */
 /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
 type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void
   ? I
   : never;
+
+/** Conditionally chooses between two types depending on whether T is a union. */
 type CheckForUnion<T, TErr, TOk> = [T] extends [UnionToIntersection<T>] ? TOk : TErr;
 
+/** Conditionally chooses a type (or void) depending on whether T is a string. */
 type CheckForStringLiteralType<T, TOk> = string extends T ? void : CheckForUnion<T, void, TOk>;
 
+type KeyOfNeverable<T> = T extends never ? never : keyof T;
+type AllKeysFromUnion<T> = keyof T | KeyOfNeverable<UnionToIntersection<T>>;
+type KeyOfOr<T, K, Default> = K extends keyof T ? T[K] : Default;
+
+/**
+ * Flatten a union of interfaces into a single interface encoding the same type.
+ *
+ * Flattens a union in such a way that:
+ * `{ a: number, b?: undefined } | { b: string, a?: undefined }`
+ * (which is the value type of `[{ a: 1 }, { b: 1 }]`)
+ * becomes `{ a: number | undefined, b: string | undefined }`.
+ *
+ * And also works for `{ a: number } | { b: string }` which maps to the same.
+ */
+type FlattenUnionOfInterfaces<T> = {
+  [K in AllKeysFromUnion<T>]: KeyOfOr<
+    T,
+    // If T always has K, just take T[K] (union of C[K] for each component C of T):
+    K,
+    // Otherwise, take the union of C[K] for each component C of T, PLUS undefined:
+    undefined | KeyOfOr<UnionToIntersection<T>, K, void>
```

Comment:
Is the following correct?
 - K is only a keyof the union T if it is present in all of the union's subtypes,
 - if we're in the default case, it means K was absent in at least one of T's subtypes
 - so UnionToIntersection<T>[K] will be never?

I believe these are the cases you describe in your comment "Unexpected test results - hopefully okay to ignore these"

Reply:
Okay nvm I see it's only a problem if in UnionToIntersection the types conflict -- that's when you'll get never. but if they're optional (undefined) then it's ok.

Reply:
Yup. `keyof ({a: number} | {b: string})` is `never` but `keyof ({a: number} & {b: string})` is `'a' | 'b'` I think. for some reason.